### PR TITLE
SI-191: add the ability to link to external projects. Review by @VladUre...

### PR DESCRIPTION
**Pull request by vigdorchik**
...che.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/doc/model/ModelFactoryTypeSupport.scala [hunk 1]
```diff
@@ -77,7 +77,8 @@ trait ModelFactoryTypeSupport {
           // ===> in such cases we have two options:
           // (0) if there's no inheritance taking place (Enum#Value) we can link to the template directly
           // (1) if we generate the doc template for Day, we can link to the correct member
-          // (2) if we don't generate the doc template, we should at least indicate the correct prefix in the tooltip
+          // (2) If the symbol comes from an externl library for which we know the documentation URL, point to it.
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryTypeSupport.scala]: Typo: `externl`.

File: src/compiler/scala/tools/nsc/doc/Settings.scala [hunk 2]
```diff
@@ -194,6 +194,12 @@ class Settings(error: String => Unit, val printMsg: String => Unit = println(_))
     "Expand all type aliases and abstract types into full template pages. (locally this can be done with the @template annotation)"
   )
 
+  val docExternalUrls = MultiStringSetting (
+    "-external-urls",
+    "externalUrl(s)",
+    "comma-separated list of package_names=doc_URL for external dependencies, where package names are ';'-separated"
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/doc/Settings.scala]: I think I used `:` as a separator for the -skip-packages command. Can you also change that to `;`?

File: src/compiler/scala/tools/nsc/doc/model/ModelFactoryTypeSupport.scala [hunk 3]
```diff
@@ -87,20 +88,30 @@ trait ModelFactoryTypeSupport {
...
+                    }.fold (
+                      // (3) if we couldn't find neither the owner nor external URL, show a tooltip with the qualified name
+                      Tooltip(name): LinkTo
+                    ) {
+                      case (_, url) =>
+                        // (2) we found external documentation to link to.
+                        LinkToExternal(name, url + "#" + name)
+                    }
+                }
             }
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryTypeSupport.scala]: I think this is being done in a couple of places -- it would be good to unite them all at one point.

File: src/compiler/scala/tools/nsc/doc/model/ModelFactoryTypeSupport.scala [hunk 4]
```diff
@@ -308,16 +319,8 @@ trait ModelFactoryTypeSupport {
...
-        case Some(typeEntity) =>
-          typeEntity
-        case None =>
-          val typeEntity = createTypeEntity
-          typeCache += aType -> typeEntity
-          typeEntity
-      }
-    } else createTypeEntity
+    if (settings.docNoPrefixes.value)
+      typeCache.getOrElseUpdate(aType, createTypeEntity)
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/doc/model/ModelFactoryTypeSupport.scala:320]: Awesome!

File: test/scaladoc/run/SI-191.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+object Test extends ScaladocModelTest {
+
+  override def code = """
+    |/** Test comment this is. */
+    |object Test {
+    |  def foo(param: Any) {}
+    |  def bar(l: List[String]) {}
+    |  def baz(d: java.util.Date) {} // Not known, even externaly
+    |}""".stripMargin
```
Comment (VladUreche) [inline test/scaladoc/run/SI-191.scala]: We should also enable linking in doc comments - look for the TODO in `MemberLookup.scala`.

```
/** [[java.lang.String A string]] is your best friend. */
def buz(d: String) {}
```

**Comment by VladUreche** [comment]
It's all good except for two cases:
1. as far as I can tell you can't link manually in the doc comment `[[java.lang.String]]`
2. as far as I can tell you can't link to members, and this can even cause incorrect links to be generated:
`trait.scala`, first scaladoc run:

```
trait T { type T }
```

`trait2.scala`, second scaladoc run:

```
trait T2 extends T { def foo: T } /* what will the link in T point to? */
```

**Comment by vigdorchik** [comment]
@VladUreche Please take another look.

**Comment by VladUreche** [comment]
PLS REBUILD ALL

**Comment by VladUreche** [comment]
LGTM, only needs a test for the doc comments

**Comment by vigdorchik** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
What's the status of this? @VladUreche: can you give LGTM once again since @vigdorchik added a new commit?

**Comment by gkossakowski** [comment]
@VladUreche: ping.

**Comment by VladUreche** [comment]
Sorry for the long response time, I just figured sweet ol' gmail considered all messages from github as spam... :((

To quote from a discussion with Eugene a couple of days ago:

> Looking from a distance, I'm not terribly happy with this solution for SI-191. Three reasons:
> 1. It only works in a handful of cases (if it's indeed a class or trait, not a type alias, type parameter or abstract type)
> 2. it can generate incorrect links
> 3. it can't link to members at all
> 
> All in all, unless we need it rushed into 2.10, I suggest we keep this for master and improve on it. I don't mean to criticize your work, but it's a delicate problem and there's a reason nobody fixed it in 5 years -- it's very very tricky to get it right. Let's move the discussion to scala-internals and see what the general opinion is.

As far as I understood Eugene is okay with delaying the patch and I proposed a solution that would fix all three problems in most of the cases -- we just need to shuffle a bunch of code around in scaladoc to get things working.

So, bottom line: let's wait until we have a more solid implementation for this.

**Comment by VladUreche** [comment]
>  Maybe we should indeed schedule a skype meeting?

Yes, that would be very useful to put us on the same page.
